protoc-gen-openapiv2/internal/genopenapi/template.go

```diff
@@ -1196,9 +1227,11 @@ func applyTemplate(p param) (*openapiSwaggerObject, error) {
 	// and create entries for all of them.
 	// Also adds custom user specified references to second map.
 	requestResponseRefs, customRefs := refMap{}, refMap{}
-	if err := renderServices(p.Services, s.Paths, p.reg, requestResponseRefs, customRefs, p.Messages); err != nil {
+	tags, err := renderServices(p.Services, s.Paths, p.reg, requestResponseRefs, customRefs, p.Messages)
+	if err != nil {
 		panic(err)
 	}
+	s.Tags = append(s.Tags, tags...)
```

Comment:
I think it'd be neater to just do the tag calculation in a separate loop here than add it to `renderServices`. Yes, it would make this O(2n) rather than O(n) but it's hardly significant, and it means we can avoid changing `renderServices`. What do you think?

Reply:
Yes, this is much cleaner and implicitly solves some of the other issues. Done.